README.md

```diff
@@ -1644,6 +1645,60 @@ variables](https://www.terraform.io/docs/configuration/environment-variables.htm
 prefix `TF_VAR_`, so one way to share a variable named `foo` between Terraform and Terragrunt is to set its value
 as the environment variable `TF_VAR_foo` and to read that value in using this `get_env()` built-in function.
 
+#### get_output
+
+`get_output(TARGET_TERRAGRUNT_CONFIG, [NAME])` returns the value of the output of the terraform module configured with
+the terragrunt config specified at `TARGET_TERRAGRUNT_CONFIG`. If provided a `NAME`, this function will return just the
+output indexed by `NAME`. Otherwise, this will return the all outputs as a map. The return type is preserved. For
+example, if the output is of type `number`, the returned value of `get_output` will also be of type `number`.
+
+Example:
+
+Suppose you had the following terragrunt live folder structure:
+
+```
+/terraform-code
+├── app1
+│   └── terragrunt.hcl
+├── tests
+│   ├── app2
+│   |   └── terragrunt.hcl
+│   └── app3
+│       └── terragrunt.hcl
+```
+
+Suppose `app1` had the following outputs:
+
+```
+output "instance_id" {}
+output "instance_name" {}
+```
+
+In `app3`, you can get each of these outputs using `get_output` in the following ways:
+
+```hcl
+inputs = {
+  instance_id_directly = get_output("${get_terragrunt_dir()}/../../app1", "instance_id")
+  instance_name_directly = get_output("${get_terragrunt_dir()}/../../app1", "instance_name")
+  instance_id_indirectly = get_output("${get_terragrunt_dir()}/../../app1")["instance_id"]
+}
+```
+
+Note that the terragrunt run configuration is inherited from the current session. What this means is that settings set
+with environment variables (such as `TERRAGRUNT_IAM_ROLE`) are automatically passed down to the execution context of
+the target terragrunt config when reading the output data.
+
+**Important**: Because `get_output` depends on the target config having been applied already, you will need to make sure
+that the specified `TARGET_TERRAGRUNT_CONFIG` is listed as a dependency in the `dependencies` block. Otherwise, this
+will fail when you run an `apply-all` since the output will be interpolated before the target config has applied.
```

Comment:
I think `get_output` should automatically count as implicit dependency, removing the need for adding the item in the `dependencies` block. This is the same way Terraform works: a reference from one resource to another creates an implicit dependency, and you use `depends_on` solely as a workaround.

In fact, `get_output` may allow us to move away from `terraform_remote_state` and manually-managed, error-prone `dependencies` entirely. Instead, each module will expose input variables for all its dependencies (e.g., `vpc_id`) and in `infrastructure-live`, you'll fill in those variables via a bunch of `get_output` calls (e.g., `vpc_id = get_output("../../vpc", "vpc_id")`). You can then call `apply-all`, and we'll automatically apply everything in the right order based on these `get_output` dependencies. 

Moreover, instead of getting a confusing error when `terraform_remote_state` of some dependency doesn't exist, the `get_output` helper can detect if a module hasn't been applied yet, and show a corresponding error message.

Reply:
Yes ideally. However, this is going to be significantly more work to implement.

In my experimentation with walking the AST in the `locals` implementation, it turns out that function calls are not super easy to reach in the parsed AST. I know this, because I noticed that those are not clear nodes. I was only hitting the vars when I was walking the tree. This was nice when I was parsing out the self references in `locals`; bad in this case where I want to parse out the function args to `get_output`.

Reply:
Not completely ruling it out: just that this should be another PR after this one.

Reply:
Do we need to walk the AST ahead of time for `get_output`? Or is it enough to record dependencies while that function is being executed as part of existing config parsing, and to then use the discovered dependency graph before executing further?

Reply:
I don't think that is possible if we are using a function, because this is rendered during the struct. We COULD build a special struct that is a placeholder result of all the `get_output` calls that is then resolved later, but I think that will get very ugly unless we restrict `get_output` to only be able to apply in `input`...

Reply:
This is now implemented with the `terragrunt_output` change because it is so much easier to partially parse individual blocks than a random function call!